src/metrics-collector/db/sqldb/sqldb_test.go

```diff
@@ -0,0 +1,377 @@
+	"metrics-collector/metrics"
+
+	"github.com/lib/pq"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/pivotal-golang/lager"
+
+	"os"
+	"time"
+)
+
+var _ = Describe("Sqldb", func() {
+	var (
+		conf       *config.DbConfig
+		db         *SQLDB
+		logger     lager.Logger
+		err        error
+		mtrcs      []*metrics.Metric
+		start      int64
+		end        int64
+		before     int64
+		appId      string
+		metricName string
+		appIds     []string
+	)
+	BeforeEach(func() {
+		logger = lager.NewLogger("sqldb-test")
+
+		dbUrl := os.Getenv("DBURL")
+		if dbUrl == "" {
+			Fail("environment variable $DBURL is not set")
+		}
+		conf = &config.DbConfig{Url: dbUrl}
+	})
+
+	Describe("NewSQLDB", func() {
+
+		JustBeforeEach(func() {
+			db, err = NewSQLDB(conf, logger)
+		})
+
+		AfterEach(func() {
+			dropMetricsTable()
+		})
+
+		Context("when db config is not correct", func() {
+			BeforeEach(func() {
+				conf.Url = "postgres://non-exist-user:non-exist-password@localhost/autoscaler?sslmode=disable"
+			})
+
+			It("should error", func() {
+				Expect(err).To(BeAssignableToTypeOf(&pq.Error{}))
+			})
+		})
+
+		Context("when db config is correct", func() {
+			Context("when there is no metrics table", func() {
+				It("should not error", func() {
+					Expect(err).To(BeNil())
+					Expect(db).NotTo(BeNil())
+				})
+
+				It("creats the table", func() {
+					Expect(hasMetricsTable()).To(BeTrue())
+				})
+			})
+
+			Context("when there is metrics table", func() {
+				BeforeEach(func() {
+					createMetricsTable()
+				})
+
+				It("should not error", func() {
+					Expect(err).To(BeNil())
+					Expect(db).NotTo(BeNil())
+				})
+
+				It("has the metrics table", func() {
+					Expect(hasMetricsTable()).To(BeTrue())
+				})
+
+			})
+		})
+	})
+
+	Describe("SaveMetric", func() {
+		BeforeEach(func() {
+			db, err = NewSQLDB(conf, logger)
+			if err != nil {
+				Fail("can not connect database")
+			}
+		})
+
+		AfterEach(func() {
+			dropMetricsTable()
+		})
+
+		Context("When insert a matric", func() {
+			BeforeEach(func() {
+				instances := []metrics.InstanceMetric{{23456312, 0, "3333"}, {23556312, 1, "6666"}}
+				metric := &metrics.Metric{
+					AppId:     "test-app-id",
+					Name:      metrics.MemoryMetricName,
+					Unit:      metrics.UnitBytes,
+					TimeStamp: 11111111,
+					Instances: instances,
+				}
+				err = db.SaveMetric(metric)
+			})
+
+			It("should not error", func() {
+				Expect(err).To(BeNil())
+			})
+
+			It("has the record in database", func() {
+				b := hasMetric("test-app-id", 11111111)
+				Expect(b).To(BeTrue())
+			})
+		})
+
+	})
+
+	Describe("RetrieveMetrics", func() {
+		BeforeEach(func() {
+			db, err = NewSQLDB(conf, logger)
+			if err != nil {
+				Fail("can not connect database")
+			}
+			start = 0
+			end = -1
+			appId = "test-app-id"
+			metricName = metrics.MemoryMetricName
+		})
+
+		AfterEach(func() {
+			dropMetricsTable()
+		})
+
+		Context("When metrics table is empty", func() {
```

Comment:
This is more basic.
It is about retrieving metrics for an app that isn't there.